private/buf/cmd/buf/command/decode/decode.go

```diff
@@ -0,0 +1,167 @@
+	sourceFlagName      = "source"
+	typeFlagName        = "type"
+	outputFlagName      = "output"
+	outputFlagShortName = "o"
+)
+
+// NewCommand returns a new Command.
+func NewCommand(
+	name string,
+	builder appflag.Builder,
+) *appcmd.Command {
+	flags := newFlags()
+	return &appcmd.Command{
+		Use:   name + " <descriptor>",
+		Short: "Decode binary descriptors with a source reference.",
+		Long: `The first argument is the serialized descriptor to decode.
+If no argument is specified, defaults to stdin.`,
+		Args: cobra.MaximumNArgs(1),
+		Run: builder.NewRunFunc(
+			func(ctx context.Context, container appflag.Container) error {
+				return run(ctx, container, flags)
+			},
+			bufcli.NewErrorInterceptor(),
+		),
+		BindFlags: flags.Bind,
+	}
+}
+
+type flags struct {
+	ErrorFormat string
+	Source      string
+	Type        string
+	Output      string
+}
+
+func newFlags() *flags {
+	return &flags{}
+}
+
+func (f *flags) Bind(flagSet *pflag.FlagSet) {
+	flagSet.StringVar(
+		&f.ErrorFormat,
+		errorFormatFlagName,
+		"text",
+		fmt.Sprintf(
+			"The format for build errors, printed to stderr. Must be one of %s.",
+			stringutil.SliceToString(bufanalysis.AllFormatStrings),
+		),
+	)
+	flagSet.StringVar(
+		&f.Source,
+		sourceFlagName,
+		"",
+		"The source that defines the serialized descriptor (e.g. buf.build/acme/weather)",
+	)
+	flagSet.StringVar(
+		&f.Type,
+		typeFlagName,
+		"",
+		"The fully-qualified type name of the serialized descriptor (e.g. acme.weather.v1.Units)",
+	)
+	flagSet.StringVarP(
+		&f.Output,
+		outputFlagName,
+		outputFlagShortName,
+		"-",
+		// TODO: If we ever support other formats (e.g. prototext), we will need
+		// to build a buffetch.DecodeOutputRefParser.
+		`The location to write the decoded result to. This is always JSON for now.`,
+	)
+}
+
+func run(
+	ctx context.Context,
+	container appflag.Container,
+	flags *flags,
+) error {
+	input, err := bufcli.GetInputValue(container, "", "-")
+	if err != nil {
+		return err
+	}
+	if err := bufcli.ValidateErrorFormatFlag(flags.ErrorFormat, errorFormatFlagName); err != nil {
+		return err
+	}
+	registryProvider, err := bufcli.NewRegistryProvider(ctx, container)
+	if err != nil {
+		return err
+	}
+	descriptorBytes, err := bufcli.NewDescriptorBytesForInput(ctx, container, registryProvider, input)
+	if err != nil {
+		return err
+	}
+	protoSource, protoType, err := parseSourceAndType(ctx, flags.Source, flags.Type)
+	if err != nil {
+		return err
+	}
+	image, err := bufcli.NewImageForSource(ctx, container, registryProvider, protoSource, flags.ErrorFormat)
+	if err != nil {
+		return err
+	}
+	message, err := bufreflect.NewMessage(ctx, image, protoType)
+	if err != nil {
+		return err
+	}
+	if err := proto.Unmarshal(descriptorBytes, message); err != nil {
+		return err
+	}
+	return bufcli.NewWireProtoEncodingWriter(
+		container.Logger(),
+	).PutMessage(
+		ctx,
+		container,
+		message,
+		flags.Output,
+	)
+}
+
+func parseSourceAndType(
```

Comment:
It'd be good to leave a lightweight comment about what's happening here (similar to the error message comment below).

It'd also be good to outline that case we discussed earlier, i.e. if someone redundantly specifies `--source buf.build/acme/weather --type buf.build/acme/weather/weather.v1.Units`.